rasa/core/actions/action_run_slot_rejections.py

```diff
@@ -0,0 +1,131 @@
+
+    async def run(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        metadata: Optional[Dict[Text, Any]] = None,
+    ) -> List[Event]:
+        """Run the predicate checks."""
+        events: List[Event] = []
+        violation = False
+        utterance = None
+        internal_error = False
+
+        dialogue_stack = DialogueStack.from_tracker(tracker)
+        top_frame = dialogue_stack.top()
+        if not isinstance(top_frame, CollectInformationPatternFlowStackFrame):
+            return []
+
+        if not top_frame.rejections:
+            return []
+
+        slot_name = top_frame.collect_information
+        slot_instance = tracker.slots.get(slot_name)
+        if slot_instance and not slot_instance.has_been_set:
+            # this is the first time the assistant asks for the slot value,
+            # therefore we skip the predicate validation because the slot
+            # value has not been provided
+            structlogger.debug(
+                "first.collect.slot.not.set",
+                slot_name=slot_name,
+                slot_value=slot_instance.value,
+            )
+            return []
+
+        slot_value = tracker.get_slot(slot_name)
+
+        current_context = dialogue_stack.current_context()
+        current_context[slot_name] = slot_value
+
+        structlogger.debug("collect.predicate.context", context=current_context)
+        document = current_context.copy()
+
+        for rejection in top_frame.rejections:
+            condition = rejection.if_
+            utterance = rejection.utter
+
+            try:
+                rendered_template = Template(condition).render(current_context)
+                predicate = Predicate(rendered_template)
+                violation = predicate.evaluate(document)
+                structlogger.debug(
+                    "collect.predicate.result",
+                    predicate=predicate.description(),
+                    violation=violation,
+                )
+            except (TypeError, Exception) as e:
+                structlogger.error(
+                    "collect.predicate.error",
+                    predicate=condition,
+                    document=document,
+                    error=str(e),
+                )
+                violation = True
+                internal_error = True
+
+            if violation:
+                break
+
+        if not violation:
+            return []
+
+        # reset slot value that was initially filled with an invalid value
+        events.append(SlotSet(top_frame.collect_information, None))
+
+        if internal_error:
+            utterance = "utter_internal_error_rasa"
+
+        if utterance is None:
+            structlogger.error(
+                "collect.rejection.missing.utter",
+                utterance=utterance,
+            )
+            return events
```

Comment:
I think the better test would be `if not isinstance(utterance, str)` so we capture all possible type violations for that field

Reply:
Addressed: https://github.com/RasaHQ/rasa/pull/12802/commits/79b1cd1224fa7df417000941b36c3619a8e12ee9